internal/dcache/rpc/server/handler.go

```diff
@@ -1468,6 +1471,117 @@ func (h *ChunkServiceHandler) Hello(ctx context.Context, req *models.HelloReques
 	}, nil
 }
 
+// Helper function to read given chunk and (optionally) the hash file.
+// It performs direct or buffered read as per the configured setting or may fallback to buffered read for
+// cases where direct read cannot be performed due to alignment restrictions.
+func readChunkAndHash(chunkPath, hashPath *string, readOffset int64, data *[]byte) (int /* read bytes */, string /* hash */, error) {
+	var fh *os.File
+	var n, fd int
+	var err error
+	var hash string
+
+	common.Assert(chunkPath != nil && len(*chunkPath) > 0)
+	common.Assert(data != nil && len(*data) > 0)
+	common.Assert(readOffset >= 0)
+
+	readLength := len(*data)
+
+	//
+	// Caller must pass data buffer aligned on FS_BLOCK_SIZE, else we have to unnecessarily perform buffered read.
+	//
+	dataAddr := unsafe.Pointer(&(*data)[0])
+	isDataBufferAligned := ((uintptr(dataAddr) % common.FS_BLOCK_SIZE) == 0)
+	common.Assert(isDataBufferAligned, uintptr(dataAddr), common.FS_BLOCK_SIZE)
```

Comment:
We can remove this assert here and at L1759, because we are already falling back to buffered IO when the data buffer in not aligned.